kafka-4592: Kafka Producer Metrics Invalid Value

huxihx opened the pull request:
Default constructor of Max assigns negative infinity as the init value.
For the record size and like, value of zero is more reasonable.

huxi-2b committed:
KAFKA-4428; Kafka does not exit if port is already bound

During Acceptor initialization, if "Address already in use" error is encountered,
the shutdown latch in each Processor is never counted down. Consequently,
the Kafka server hangs when `Processor.shutdown` is called.

Author: huxi <huxi@zhenrongbao.com>
Author: amethystic <huxi_2b@hotmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2156 from amethystic/kafka-4428_Kafka_noexit_for_port_already_use

vahidhashemian committed:
MINOR: Sync up 'kafka-run-class.bat' with 'kafka-run-class.sh'

Some of the recent changes to `kafka-run-class.sh` have not been applied to `kafka-run-class.bat`.
These recent changes include setting proper streams or connect classpaths. So any streams or connect use case that leverages `kafka-run-class.bat` would fail with an error like
```
Error: Could not find or load main class org.apache.kafka.streams.???
```

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2238 from vahidhashemian/minor/sync_up_kafka-run-class.bat

eribeiro committed:
MINOR: Mx4jLoader always returns false even if mx4j is loaded & started

Mx4jLoader.scala should explicitly `return true` if the class is successfully loaded and started, otherwise it will return false even if the class is loaded.

Author: Edward Ribeiro <edward.ribeiro@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2295 from eribeiro/mx4jloader-bug

LoneRifle committed:
MINOR: Rephrase Javadoc summary for ConsumerRecord

The original Javadoc description for `ConsumerRecord` is slightly confusing in that it can be read in a way such that an object is a key value pair received from Kafka, but (only) consists of the metadata associated with the record. This PR makes it clearer that the metadata is _included_ with the record, and moves the comma so that the phrase "topic name and partition number" in the sentence is more closely associated with the phrase "from which the record is being received".

Author: LoneRifle <LoneRifle@users.noreply.github.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2290 from LoneRifle/patch-1

huxi-2b committed:
KAFKA-4351; MirrorMaker with new consumer should support comma-separated regex

This makes it consistent with MirrorMaker with the old consumer.

Author: huxi <huxi@zhenrongbao.com>
Author: amethystic <huxi_2b@hotmail.com>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2072 from amethystic/kafka-4351_Regex_behavior_change_for_new_consumer

vahidhashemian committed:
KAFKA-4480: Report an error in 'kafka-configs' command if the config to be removed does not exist

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2218 from vahidhashemian/KAFKA-4480

himani1 committed:
MINOR: Scala code readability improvements

Author: Himani Arora <1himani.arora@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2297 from himani1/refactored_code

baluchicken committed:
KAFKA-4318; Migrate ProducerSendTest to the new consumer

Author: Balint Molnar <balintmolnar91@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2083 from baluchicken/KAFKA-4318

dguy committed:
KAFKA-4561; Ordering of operations in StreamThread.shutdownTasksAndState may void at-least-once guarantees

In `shutdownTasksAndState` and `suspendTasksAndState` we commit offsets BEFORE we flush any state. This is wrong as if an exception occurs during a flush, we may violate the at-least-once guarantees, that is we would have committed some offsets but NOT sent the processed data on to other Sinks.
Also during suspend and shutdown, we should try and complete all tasks even when exceptions occur. We should just keep track of the exception and rethrow it at the end if necessary. This helps with ensuring that StateStores etc are closed.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2281 from dguy/kafka-4561

Commit:
KAFKA-2434: Remove identical topic subscription constraint for roundrobin strategy in old consumer API

Author: Andrew Olson <aolson1@cerner.com>

Reviewers: Onur Karaman

Closes #145 from noslowerdna/KAFKA-2434

ijuma committed:
KAFKA-4528: Fix failure in ProducerTest.testAsyncSendCanCorrectlyFailWithTimeout

I was able to reproduce the failure in less than 10 runs before the change. With the change,
the test passed 70 times consecutively.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang

Closes #2298 from ijuma/kafka-4528-fix-test-async-send-timeout

ewencp committed:
KAFKA-4404; Add javadocs to document core Connect types, especially that integer types are signed

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Konstantine Karantasis <konstantine@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2296 from ewencp/kafka-4404-document-connect-signed-integer-types

lindong28 committed:
KAFKA-4429; Consumer lag metric should be zero if FetchResponse is empty

Author: Dong Lin <lindong28@gmail.com>

Reviewers: Ewen Cheslack-Postava <me@ewencp.org>, Jason Gustafson <jason@confluent.io>

Closes #2155 from lindong28/KAFKA-4429

fpj committed:
KAFKA-4261: Provide debug option in vagrant-up.sh

Author: fpj <fpj@apache.org>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1981 from fpj/vagrant-debug-option

ijuma committed:
MINOR: Improvements to Record related classes (part 1)

Jason recently cleaned things up significantly by consolidating the Message/Record classes
into the common Java code in the clients module. While reviewing that, I noticed a few things
that could be improved a little more. To make reviewing easier, there will be multiple PRs.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <me@ewencp.org>, Jason Gustafson <jason@confluent.io>

Closes #2271 from ijuma/records-minor-fixes

vahidhashemian committed:
KAFKA-4584: Fail the 'kafka-configs' command if the config to be removed does not exist

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2300 from vahidhashemian/KAFKA-4584

huxi-2b committed:
KAFKA-4434; KafkaProducer configuration is logged twice

ProducerConfig calls AbstractConfig.init where does the logs. KafkaProducer init will inovoke ProducerConfig.init twice that leads to logging twice.

Author: huxi <huxi@zhenrongbao.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2307 from amethystic/kafka-4434_Kafkaproducer_log_twice

rajinisivaram committed:
KAFKA-4583; Fix KafkaConsumerTest.testGracefulClose transient failure

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2309 from rajinisivaram/KAFKA-4583

cmccabe committed:
KAFKA-4508; System test that runs client against older versions of the broker

In reality, we’ll only test older brokers after KAFKA-4462 is fully implemented.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2263 from cmccabe/KAFKA-4508

Commit:
MINOR: Remove stray comma from PartitionInfo.toString

Removed the extra ',' character while printing the replicas / in-sync replicas
array.

Author: Kamal <kamal@nmsworks.co.in>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2306 from Kamal15/trunk

Commit:
KAFKA-4523; Fix crash during shutdown due to group coordinator attempting to write to a closed log

Shut down the group coordinator before shutting down the log manager to
ensure that any delayed operations are completed before the logs are closed.

Author: steve <sniemitz@twitter.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2311 from steveniemitz/KAFKA-4523

huxi-2b committed:
kafka-4592: Kafka Producer Metrics Invalid Value

Default constructor of Max assigns negative infinity as the init value.
For the record size and like, value of zero is more reasonable.

huxihx commented:
@ijuma @hachikuji Please review the PR. Thanks.

The reason why I added a boolean flag in Max is consider the possibility of all negative values in the metrics, although all the **-max metrics are now non-negative. Does it make sense?

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/Max.java:
```diff
@@ -21,8 +21,15 @@
+
     public Max() {
-        super(Double.NEGATIVE_INFINITY);
+        this(false);
```
Having this as the default doesn't seem right. Not to mention backwards compatibility issues.

huxihx replied:
agreed, but we have to change all outer code which created the Max instance.  Many places. Since all the metrics are now non-negative, that's why I change the default behavior that indeed breaks the backwards compatibility rule. Which one do you prefer?

ijuma replied:
I think it's OK to change the calling code.

huxihx replied:
Another thing to confirm. Since this jira issue only mentions records-lag-max and record-size-max, should we correct the code for these two metrics or all the metrics that uses Max?

ijuma replied:
All the metrics that are buggy.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/Max.java:
```diff
@@ -21,8 +21,15 @@
+        this(false);
+    }
+
+    public Max(boolean whetherCanBeNegative) {
```
A more general approach would be to allow one to pass `initialValue` or something like that.

huxihx replied:
Same reason as comment above.

ijuma replied:
It seems OK to say `Max(0)` instead of `Max(true)`.

ijuma reviewed (commented):
Thanks for the PR, left a couple of comments.

huxi-2b committed:
kafka-4592: Kafka Producer Metrics Invalid Value

1. changed Max constructor from Max(boolean) to Max(double)
2. refined all metrics that use Max to have them invoke Max(0.0) instead
of Max(0)

huxihx commented:
@ijuma Thanks for the comments. Please review again.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/Max.java:
```diff
@@ -21,8 +21,16 @@
+
+    public Max(double initValue) {
+        super(initValue);
+        this.initValue = initValue;
```
You should access `initialValue` from the superclass instead of duplicating it. And while you're at it, make `SampledStat.initialValue` and `SampledStat.samples` final.

mjsax committed:
MINOR: Update JavaDoc for KTable helper interfaces

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Eno Thereska, Guozhang Wang

Closes #2315 from mjsax/javaDocImprovements2

shikhar committed:
KAFKA-4575: ensure topic created before starting sink for ConnectDistributedTest.test_pause_resume_sink

Otherwise in this test the sink task goes through the pause/resume cycle with 0 assigned partitions, since the default metadata refresh interval is quite long

Author: Shikhar Bhushan <shikhar@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2313 from shikhar/kafka-4575

huxi-2b committed:
kafka-4592: Kafka Producer Metrics Invalid Value

removed duplicated variable used in Max

huxihx commented:
@ijuma Thanks for reminding. Please review again.

benstopford committed:
MINOR: Add unit tests to the ReassignPartitionsCommand

Adds a bunch of tests to unit tests to the assignment command.
Moves the Rack aware test into its own class as it makes use of ZooKeeperTestHarness and slows everything else down.

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1950 from benstopford/os-rebalance-extra-unit-testing

mjsax committed:
MINOR: update JavaDoc for simple helper interfaces of KStream and KTable operators

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2321 from mjsax/javaDocImprovements3

dguy committed:
KAFKA-3452: Support session windows

Add support for SessionWindows based on design detailed in https://cwiki.apache.org/confluence/display/KAFKA/KIP-94+Session+Windows.
This includes refactoring of the RocksDBWindowStore such that functionality common with the RocksDBSessionStore isn't duplicated.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Eno Thereska <eno.thereska@gmail.com>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

Closes #2166 from dguy/kafka-3452-session-merge

mjsax committed:
MINOR: cleanup Kafka Streams exception classes

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2314 from mjsax/javaDocImprovements

raghavgautam committed:
KAFKA-4465: Create docker image and scripts for running tests locally

Author: Raghav Kumar Gautam <raghav@apache.org>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Sriharsha Chintalapani <harsha@hortonworks.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2197 from raghavgautam/trunk

yaojuncn committed:
KAFKA-4402: make the KafkaProducer true round robin per topic

Author: yaojuncn <yaojuncn@users.noreply.github.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Vahid Hashemian <vahidhashemian@us.ibm.com>, Konstantin <konstantin@tubemogul.com>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2128 from yaojuncn/KAFKA-4402-client-producer-round-robin-fix

dguy committed:
KAFKA-4551: StreamsSmokeTest.test_streams intermittent failure

Remove use of TestTimestampExtractor as it causes the logs to roll and segments get deleted.
Remove the wcnt example as it is dependent on the TestTimestampExtractor - windowed counting is covered elsewhere.
Change all aggregate operations to use TimeWindow as use of UnlimitedWindow was causing logs to roll and segments being deleted.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang, Eno Thereska

Closes #2319 from dguy/smoke-test

mjsax committed:
MINOR: some public JavaDoc cleanup

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Eno Thereska, Guozhang Wang

Closes #2332 from mjsax/javaDocImprovements5

guozhangwang committed:
MINOR: Maybe decorate inner topics for SourceNode

When creating the source node in TopologyBuilder, we need to decorate its input topics if they are inner (i.e. repartition) topics with the prefix.

Also did some minor cleanup in the printing function for better visualization in debugging.

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Eno Thereska, Damian Guy, Eno Thereska, Jun Rao

Closes #2320 from guozhangwang/KMinor-source-topic-fix

reftel committed:
TRIVIAL: Fix spelling of log message

Author: Magnus Reftel <magnus.reftel@skatteetaten.no>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2335 from reftel/feature/triggering

nixsticks committed:
MINOR: Fix small error in javadoc for persistent Stores

Closed the last PR I made for this because I accidentally borked it with my other PR. Small error; I figure this is from copy-pasting the above doc

Author: Nikki Thean <nthean@etsy.com>

Reviewers: Matthias J. Sax <matthias@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2336 from nixsticks/trunk

enothereska committed:
HOTFIX: Convert exception introduced in KAFKA-3637 to warning since clearly it is happening.

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2283 from enothereska/hotfix-state-transition-warn

benstopford committed:
MINOR: Add Replication Quotas Test Rig

This test rig lives in the other.kafka package so isn't part of our standard tests. It provides a convenient mechanism for measuring throttling performance over time. Measurements for each experiment are charted and presented to the user in an html file. The output looks like this:

**Experiment4**
- BrokerCount: 25
- PartitionCount: 100
- Throttle: 4,000,000 B/s
- MsgCount: 1,000
- MsgSize: 100,000
- TargetBytesPerBrokerMB: 400

![image](https://cloud.githubusercontent.com/assets/1297498/19070450/3251bc52-8a23-11e6-88fe-94de6b9147c2.png)
![image](https://cloud.githubusercontent.com/assets/1297498/19070467/4c19f38e-8a23-11e6-986a-ba19d16819ca.png)

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1957 from benstopford/throttling-test-rig

huxihx commented:
@ijuma please kindly review the PR. Thanks.

mimaison committed:
MINOR: Remove unnecessary semi-colons

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2326 from mimaison/minor-fixes

rajinisivaram committed:
KAFKA-3751; SASL/SCRAM implementation

Implementation of KIP-84: https://cwiki.apache.org/confluence/display/KAFKA/KIP-84%3A+Support+SASL+SCRAM+mechanisms

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2086 from rajinisivaram/KAFKA-3751

dguy committed:
MINOR: rename SessionStore.findSessionsToMerge to findSessions

Rename `SessionStore.findSessionsToMerge` to `findSessions`

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2339 from dguy/minor-findsession-rename

Commit:
MINOR: Replacing for with foreach loop in stream test classes

Author: Prabhat Kashyap <prabhat.kashyap@knoldus.in>

Reviewers: Ismael Juma, Damian Guy, Guozhang Wang

Closes #2305 from PKOfficial/code-refactor

xvrl committed:
KAFKA-4481: relax streams api type contraints

Make appropriate methods contravariant in key and value types.

Author: Xavier Léauté <xavier@confluent.io>

Reviewers: Damian Guy <damian.guy@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

Closes #2205 from xvrl/streams-contravariance

Commit:
KAFKA-4060: Remove zk client dependency in kafka streams

dguy guozhangwang This is a new PR for KAFKA-4060.

Author: Hojjat Jafarpour <hojjat@Hojjat-Jafarpours-MBP.local>
Author: Hojjat Jafarpour <hojjat@HojjatJpoursMBP.attlocal.net>

Reviewers: Damian Guy, Matthias J. Sax, Isamel Juma, Guozhang Wang

Closes #1884 from hjafarpour/KAFKA-4060-Remove-ZkClient-dependency-in-Kafka-Streams-new

jozanek committed:
MINOR: Update rocksDB dependency to 5.0.1

Author: jozi-k <jozef.koval@protonmail.ch>

Reviewers: Ismael Juma, Guozhang Wang

Closes #2292 from jozi-k/update-rocksdb-4.13.5

cmccabe committed:
KAFKA-4507; Clients should support older brokers (KIP-97)

The client should send older versions of requests to the broker if necessary.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2264 from cmccabe/KAFKA-4507

Commit:
KAFKA-3715: Add granular metrics to Kafka Streams and add hierarhical logging levels to Metrics

Kafka Streams: add granular metrics per node and per task, also expose ability to register non latency metrics in StreamsMetrics
Also added different recording levels to Metrics.

This is joint contribution from Eno Thereska and Aarti Gupta.

from https://github.com/apache/kafka/pull/1362#issuecomment-218326690-------
We can consider adding metrics for process / punctuate / commit rate at the granularity of each processor node in addition to the global rate mentioned above. This is very helpful in debugging.

We can consider adding rate / total cumulated metrics for context.forward indicating how many records were forwarded downstream from this processor node as well. This is helpful in debugging.

We can consider adding metrics for each stream partition timestamp.
This is helpful in debugging.
## Besides the latency metrics, we can also add throughput latency in terms of source records consumed.

More discussions here https://issues.apache.org/jira/browse/KAFKA-3715, KIP-104, KIP-105

Author: Eno Thereska <eno@confluent.io>
Author: Aarti Gupta <aartiguptaa@gmail.com>

Reviewers: Greg Fodor, Ismael Juma, Damian Guy, Guozhang Wang

Closes #1446 from aartigupta/trunk

edoardocomar committed:
KAFKA-4180; Support clients with different authentication credentials in the same JVM

Changed caching in LoginManager to allow one LoginManager per client
JAAS configuration.

Added test to End2EndAuthorization for SASL Plain and GSSAPI with two
consumers with different credentials.

Developed with mimaison.

Author: Edoardo Comar <ecomar@uk.ibm.com>

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2293 from edoardocomar/KAFKA-4180d

rajinisivaram committed:
KAFKA-4426; Add close with timeout for KafkaConsumer (KIP-102)

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Apurva Mehta <apurva.1618@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2285 from rajinisivaram/KAFKA-4426

huxihx commented:
@ijuma Do you think the patch is okay for this PR?

bbejeck committed:
KAFKA-4114: Allow different offset reset strategies

mjsax

Here's my first pass at finer grained auto offset reset strategies.

I've left TODO comments about whether we want to consider adding this to `KGroupedTable.aggregate` and `KStreamImpl` when re-partitioning a source.

Author: bbejeck <bbejeck@gmail.com>

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

Closes #2007 from bbejeck/KAFKA-4114_allow_different_offset_reset_strategies

mjsax committed:
KAFKA-3452; Follow-up: Add SessionWindows

 - TimeWindows represent half-open time intervals while SessionWindows represent closed time intervals

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy <damian.guy@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

Closes #2342 from mjsax/kafka-3452-session-window-follow-up

Commit:
KAFKA-4603: Fixed the argument of shell in doc

    KAFKA-4603 the argument of shell in doc wrong and command parsed error
    In "7.6.2 Migrating clusters" of document security.html, the argument "--zookeeper.connection" of shell "zookeeper-security-migrat.sh"   is wrong  and the using of OptionParser is not correct
    This patch corrected the doc and changed the OptionParser constructor

Author: auroraxlh <xin.lihua1@zte.com.cn>

Reviewers: Xi Hu <huxi_2b@hotmail.com>, Guozhang Wang <wangguoz@gmail.com>

Closes #2322 from auroraxlh/my-issue

dasl- committed:
MINOR: Fix error in design docs

Author: dasl <dasl-@users.noreply.github.com>

Reviewers: Gwen Shapira

Closes #2357 from dasl-/trunk

dguy committed:
KAFKA-4490: Add Global Table support to Kafka Streams

Add Global Tables to KafkaStreams. Global Tables are fully replicated once-per instance of KafkaStreams. A single thread is used to update them. They can be used to join with KStreams, KTables, and other GlobalKTables. When participating in a join a GlobalKTable is only ever used to perform a lookup, i.e., it will never cause data to be forwarded to downstream processor nodes.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Eno Thereska, Guozhang Wang

Closes #2244 from dguy/global-tables

huxi-2b committed:
Merge branch 'trunk' into KAFKA-4592_InfinityValue_For_metrics

huxihx commented:
@ijuma Please kindly review the PR. Thanks.

huxihx commented:
@ijuma could you please take some time to have a look at it. Many Thanks.

ijuma commented:
Sorry for the delay, I won't be able to look at this until Monday.

ijuma commented:
@amethystic, a few changes in trunk mean that you need to merge trunk to your branch and then cherry-pick the following commit:

https://github.com/ijuma/kafka/commit/9911bf96e2288735c1338207634214dbf6f2bdf0

ijuma commented:
I asked @junrao about this and he thinks that using `0` may be more confusing as it's a valid value. So, it seems like we need more discussion on this one.

huxihx commented:
In my opinion, `0` is a reasonable default value for such metrics. Anyway please have me updated if any further discussion is needed.

huxihx commented:
@ijuma What's the status for this PR? Any further discussion I had missed?
